var documenterSearchIndex = {"docs":
[{"location":"#DNASeqFISHChromosomeAssignment.jl","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.jl","text":"","category":"section"},{"location":"","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.jl","text":"This package implements the Longest Disjoint Paths algorithm for identifying chromosome strands in DNA SeqFISH+ data. The intuition behind this method DNA SeqFISH+ is like imaging beads on an invisible string, where the loci we measure are the beads and the genomic DNA is the string. We know the spatial and genomic coordinates of each locus we find, so we can infer that DNA strands connect loci that are close both spatially and genomically.","category":"page"},{"location":"","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.jl","text":"However, we do not try to model the branching nature of replicating chromosomes. To correctly classify replicated loci on chromosome pairs that are spatially distinct, we combine the LDP chromosome identification method with DBSCAN.","category":"page"},{"location":"","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.jl","text":"assign_chromosomes\r\nChromSepParams\r\nChromSepParams()\r\nset_r_dbscan\r\nset_r_ldp\r\nset_r_ldp_nbr\r\nset_sigma\r\nset_min_size\r\nset_dbscan_min_nbrs\r\nset_minProp_unique","category":"page"},{"location":"#DNASeqFISHChromosomeAssignment.assign_chromosomes","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.assign_chromosomes","text":"assign_chromosomes(pnts :: DataFrame, params :: ChromSepParams, optimizer)\n\nThis function assigns genomic loci from DNA seqFISH+ data to indiviual chromosomes using DBSCAN and Longest Disjoint Paths (LDP). First, genomic loci are clustering using DBSCAN. Then DBSCAN clusters that contain a proportion of unique genomic loci less than the minpropunique parameter set in the params object and whose average spatial distance between subsequent spatial loci is greater than the r_ldp parameter (also et in the params object) will be split, using the LDP algorithm. If these conditions are not met, then we find one LDP that is a subset of the cluster.\n\nLDP works by finding the longest one or two weighted paths of loci of increasing genomic coordinate within the a DBSCAN cluster. More specifically, it builds a directed acyclic graph where every locus has an out going edge to genomically down stream loci within the search radius. Edges are weighted by:\n\nw = frac1g_p - g_c  e^-fracx_p - x_c^22Ïƒ^2\n\nwhere x denotes a vector of spatial coordinates, g denotes genomic coordinates, and p and c subscripts denote parent and child nodes respectively. LDP then uses Integer Programming to find the one or two disjoint paths that maximize the sum of the weights of all edges in the paths.\n\nArguments\n\ndf: Takes a DataFrame (pnts) of decoded DNA seqFISH data where each row describes a locus. Must include columns:\n\n- fov : the field of view in which locus was imaged\n- cellID : the cell in which the locus was imaged\n- chrom : chromosome on which locus is located\n- x : x spatial coordinate of locus\n- y : y spatial coordinate of locus\n- z : z spatial coordinate of locus\n- g : genomic coordinate of locus\n\nprms : ChromSepParams object containing parameters.\noptimizer : optional - JuMP compatible optimizer for performing Longest Disjoint Paths optimization. Uses GLPK by default.\n\nreturns a copy of the input points sorted by fov, cellID, chrom, and g with a new columns:  \tdbscan_allele : DBSCAN clusters of each locus \tdbscan_ldp_allele : ldp subclusters of each dbscan cluster \tdbscan_ldp_nbr_allele : dbscan clusters that are not split, and expands split ldp clusters by assigning ldp neighbors to their cluster.\n\nallele = -1 means that a locus was not assigned to a chromosome. allele > 0 indicates the allele number of the chromosome that the locus was assigned to.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.ChromSepParams","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.ChromSepParams","text":"mutable struct ChromSepParams(\n    r_dbscan :: Real\n    r_ldp :: Real\n    r_ldp_nbr :: Real\n    sigma :: Real\n    min_size :: Int64\n    dbscan_min_nbrs :: Int64\n    min_prop_unique :: Real\n    )\n\nr_dbscan : the DBSCAN search radius\nr_ldp : the longest disjoint path search radius\nr_ldp_nbr : the search radius to add neighboring loci to LDP clusters after separating a DBSCAN cluster\nsigma : defines the weight function for how much to penalize grouping adjacent genomic loci onto the the same chromosome based on their spatial distance from each other.\nmin_size : the minimum acceptable number of loci on a strand.\ndbscan_min_pnts : the minimum allowed number of points for a DBSCAN cluster. DBSCAN clusters with fewer points are discarded.\nmin_prop_unique : the minimum proportion of unique loci a DBSCAN cluster is allowed to have before considering splitting it into longest disjoint paths.\n\n\n\n\n\n","category":"type"},{"location":"#DNASeqFISHChromosomeAssignment.ChromSepParams-Tuple{}","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.ChromSepParams","text":"ChromSepParams()\n\nConstructor for ChromSepParams object using the default parameter values. The user can set the parameter values using parameter setting methods.\n\n\n\n\n\n","category":"method"},{"location":"#DNASeqFISHChromosomeAssignment.set_r_dbscan","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_r_dbscan","text":"set_r_dbscan(p :: ChromSepParams, r :: Real)\n\nSet the DBSCAN search radius for clustering loci.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_r_ldp","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_r_ldp","text":"set_r_ldp(p :: ChromSepParams, r :: Real)\n\nSet the longest disjoint path search radius for clustering loci.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_r_ldp_nbr","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_r_ldp_nbr","text":"set_r_ldp_nbr(p :: ChromSepParams, r :: Real)\n\nSet the search radius for clustering loci near LDPs after separating DBSCAN clusters with LDPs.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_sigma","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_sigma","text":"set_sigma(p :: ChromSepParams, s :: Real)\n\nSet the sigma for the spatial distance penalty in the longest disjoint paths optimization.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_min_size","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_min_size","text":"set_min_size(p :: ChromSepParams, min_size :: Real)\n\nSet the minimum size of DBSCAN clusters and minimum length of the longest disjoint paths.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_dbscan_min_nbrs","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_dbscan_min_nbrs","text":"set_dbscan_min_nbrs(p :: ChromSepParams, mns :: Int64)\n\nSet the minimum number of neighbors a locus must have within the DBSCAN radius to be incuded in a DBSCAN cluster.\n\n\n\n\n\n","category":"function"},{"location":"#DNASeqFISHChromosomeAssignment.set_minProp_unique","page":"DNASeqFISHChromosomeAssignment.jl","title":"DNASeqFISHChromosomeAssignment.set_minProp_unique","text":"set_minProp_unique(p :: ChromSepParams, mpu :: Real)\n\nIf the proportion of unique loci in a DBSCAN cluster is less than this, the algorithm will then check whether the  average spatial distance between loci of subsequent genomic coordinates is greater than r_ldp. If so, then the DBSCAN cluster is split into to Longest Disjoint paths.\n\n\n\n\n\n","category":"function"}]
}
